### PR TITLE
Bug 1866318: [RHOCS Usability Study][Dashboard] Users found it difficult to navigate to the OCS dashboard

### PR DESCRIPTION
**Pull request by SanjalKatiyar**
Migrated OCS (persistent-storage and object-service) dashboards from `Home --> Overview` to `Storage --> Overview` section.

**Before:**
Screenshot from 2021-04-06 17-41-08

**After:**
Screenshot from 2021-04-06 17-14-12

### CODE REVIEW

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx [hunk 1]
```diff
@@ -6,69 +6,20 @@ import { Helmet } from 'react-helmet';
 import { useTranslation } from 'react-i18next';
 
 import { ClusterDashboard } from './cluster-dashboard/cluster-dashboard';
-import { HorizontalNav, PageHeading, LoadingBox, Page, AsyncComponent } from '../../utils';
-import Dashboard from '@console/shared/src/components/dashboard/Dashboard';
```
Comment (bipuladh) [inline frontend/public/components/dashboard/dashboards-page/dashboards.tsx:10]: Is OCS the only one using this? Even if so it might be useful for others.

File: frontend/packages/ceph-storage-plugin/src/plugin.ts [hunk 2]
```diff
@@ -653,6 +655,31 @@ const plugin: Plugin<ConsumedExtensions> = [
       required: [NOOBAA_FLAG],
     },
   },
+  {
+    type: 'NavItem/Href',
+    properties: {
+      id: 'ocsdashboards',
+      section: 'storage',
+      insertBefore: 'networkpolicies',
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/plugin.ts:663]:
```suggestion
      insertBefore: 'persistentvolumes,
```

File: frontend/packages/ceph-storage-plugin/src/plugin.ts [hunk 3]
```diff
@@ -653,6 +655,31 @@ const plugin: Plugin<ConsumedExtensions> = [
...
+      componentProps: {
+        name: '%ceph-storage-plugin~Overview%',
+        href: '/ocs-dashboards',
+      },
+    },
+    flags: {
+      required: [NOOBAA_FLAG, OCS_FLAG],
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/plugin.ts:670]:
```suggestion
      required: [OCS_FLAG],
```

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx [hunk 4]
```diff
@@ -6,69 +6,20 @@ import { Helmet } from 'react-helmet';
...
-        <Dashboard>
-          <DashboardGrid
-            mainCards={getCardsOnPosition(tabCards, GridPosition.MAIN)}
-            leftCards={getCardsOnPosition(tabCards, GridPosition.LEFT)}
-            rightCards={getCardsOnPosition(tabCards, GridPosition.RIGHT)}
-          />
-        </Dashboard>
-      ),
-    };
-  });
```
Comment (bipuladh) [inline frontend/public/components/dashboard/dashboards-page/dashboards.tsx:48]: Export these and use it in OCS package.

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx [hunk 5]
```diff
@@ -6,69 +6,20 @@ import { Helmet } from 'react-helmet';
...
-    () => [
-      {
-        href: '',
-        name: t('dashboard~Cluster'),
-        component: ClusterDashboard,
-      },
-      ...pluginPages,
-    ],
-    [pluginPages, t],
-  );
```
Comment (bipuladh) [inline frontend/public/components/dashboard/dashboards-page/dashboards.tsx:71]: Again others might find this useful for adding the dashboard tab next to overview?

**Review by bipuladh — CHANGES_REQUESTED** [review]

**Comment by bipuladh** [comment]
Please add some scrshots as well.

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 2 above)
Comment (SanjalKatiyar) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts:663]: Won't work with "persistentvolumes"... it is referring the resource by its "id" (check `admin-nav.tsx --> NavSection (Storage) --> Resource (persistentvolumes)`; "id" used here is  "networkpolicies") https://github.com/openshift/console/blob/8b9c03c10635f3ea310b74deeb0576c56359e735/frontend/public/components/nav/admin-nav.tsx#L211
Should I change the "id" corresponding to "persistentvolumes" in admin-nav.tsx page??

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 3 above)
Comment (SanjalKatiyar) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts:670]: ack...

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 1 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:10]: 1) Yes, only OCS is using `Page & AsyncComponent` from 'utils'. Though `HorizontalNav, PageHeading & LoadingBox` from the `utils` are being used by cluster-dashborad as well and are imported below.
2) `Dashboard` component is also used by both OCS and cluster-dashboard, but for cluster-dashboard this `Dashboard` component is imported in different file named `cluster-dashboard.tsx`.
Please have a look and if you want I can merge  `cluster-dashboard.tsx` and `dashboards.tsx` into one file.

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 5 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:71]: Should I leave the code as a comment in the file then??

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 1 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:10]: got it... u want me to export these functions from here and import them in `ocs-dashboards.tsx` file that I created, right??

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 4 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:48]: alright... i will keep this code here only and import these function in `ocs-dashboards.tsx` file that i created under `ceph-storage-plugin`...

Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:48]: ack... I will export these functions from here and import them in ocs-dashboards.tsx file that I created under `ceph-storage-plugin` directory...

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 5 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:71]: ack...

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  DashboardsCard,
+  DashboardsTab,
+  isDashboardsCard,
+  isDashboardsTab,
+} from '@console/plugin-sdk';
+import {
+  getPluginTabPages,
+  mapStateToProps,
+  DashboardsPageProps,
+} from '../../../../../public/components/dashboard/dashboards-page/dashboards';
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:17]:
```suggestion
} from '@console/public/components/dashboard/dashboards-page/dashboards';
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import { HorizontalNav, PageHeading, LoadingBox } from '../../../../../public/components/utils';
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:18]:
```suggestion
import { HorizontalNav, PageHeading, LoadingBox } from '@console/public/components/utils';
```

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 2 above)
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts:663]: oh okay, wow yeah please change it there as well. It's a mistake.

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  const title = t('ceph-storage-plugin~OpenShift Container Storage Overview');
+  const tabExtensions = useExtensions<DashboardsTab>(isDashboardsTab);
+  const cardExtensions = useExtensions<DashboardsCard>(isDashboardsCard);
+
+  const pluginPages = React.useMemo(
+    () => getPluginTabPages(tabExtensions, cardExtensions, 'persistent-storage'),
+    [tabExtensions, cardExtensions],
+  );
+
+  const allPages = React.useMemo(() => [...pluginPages], [pluginPages]);
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:31]: this is unnecessary, just pass the `pluginPages` to `HorizontalNav`

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  mapStateToProps,
+  DashboardsPageProps,
+} from '@console/internal/components/dashboard/dashboards-page/dashboards';
+import { HorizontalNav, PageHeading, LoadingBox } from '@console/internal/components/utils';
+
+const OCSDashboardsPage: React.FC<DashboardsPageProps> = ({ match, kindsInFlight, k8sModels }) => {
+  const { t } = useTranslation();
+  const title = t('ceph-storage-plugin~OpenShift Container Storage Overview');
+  const tabExtensions = useExtensions<DashboardsTab>(isDashboardsTab);
+  const cardExtensions = useExtensions<DashboardsCard>(isDashboardsCard);
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:24]: You cannot use these extensions. They are meant for Cluster overview. If anybody will want to add a new tab next tu cluster overview it will appear here.

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx [hunk 10]
```diff
@@ -50,25 +51,14 @@ const getPluginTabPages = (tabs: DashboardsTab[], cards: DashboardsCard[]): Page
 const DashboardsPage_: React.FC<DashboardsPageProps> = ({ match, kindsInFlight, k8sModels }) => {
   const { t } = useTranslation();
   const title = t('dashboard~Overview');
-  const tabExtensions = useExtensions<DashboardsTab>(isDashboardsTab);
-  const cardExtensions = useExtensions<DashboardsCard>(isDashboardsCard);
```
Comment (rawagner) [inline frontend/public/components/dashboard/dashboards-page/dashboards.tsx:54]: You're removing a chance to add a new tab next to cluster overview completely. We still want that

**Review by rawagner — CHANGES_REQUESTED** [review]

**Comment by rawagner** [comment]
cc @andybraren

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 10 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx]: Thanks for your reviews. I have made the changes, PTAL.

File: frontend/packages/ceph-storage-plugin/src/plugin.ts [hunk 11]
```diff
@@ -653,6 +620,32 @@ const plugin: Plugin<ConsumedExtensions> = [
...
   },
+  {
+    type: 'NavItem/Href',
+    properties: {
+      id: 'ocsdashboards',
+      section: 'storage',
+      insertBefore: 'persistentvolumes',
+      componentProps: {
+        // t('ceph-storage-plugin~Overview')
+        name: '%ceph-storage-plugin~Overview%',
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/plugin.ts:631]: I think this should be `OCS Overview` calling it Overview means we are giving an overall storage overview which we are not giving.

@rawagner @bipuladh @yuvalgalanti Opinions?

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx [hunk 12]
```diff
@@ -50,25 +47,24 @@ const getPluginTabPages = (tabs: DashboardsTab[], cards: DashboardsCard[]): Page
 const DashboardsPage_: React.FC<DashboardsPageProps> = ({ match, kindsInFlight, k8sModels }) => {
   const { t } = useTranslation();
   const title = t('dashboard~Overview');
+
+  /* To add a new tab next to cluster overview */
```
Comment (rawagner) [inline frontend/public/components/dashboard/dashboards-page/dashboards.tsx:51]: This cannot be commented out. We want to keep the ability to add new tabs to cluster overview

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+      />
+    </Dashboard>
+  );
+};
+
+const OCSDashboardsPage: React.FC<DashboardsPageProps> = ({ match, kindsInFlight, k8sModels }) => {
+  const { t } = useTranslation();
+  const title = t('ceph-storage-plugin~OpenShift Container Storage Overview');
+  const isOcsIndependent = useFlag(OCS_INDEPENDENT_FLAG);
+  const cardExtensions = useExtensions<DashboardsCard>(isDashboardsCard);
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:36]: You're still using extensions meant for cluster overview. You need to create your own.

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 11 above)
Comment (rawagner) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts:631]: +1

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx (hunk 13 above)
Comment (rawagner) [reply frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:36]: or you could use the same but you need a way to identify to which dashboard the extension should be added. Maybe some URL matching

Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:36]:
```suggestion
  const cardExtensions = useExtensions<DashboardsCard>((e) => isDashboardsCard && (['independent-dashboard', 'persistent-storage', 'object-service'].includes(e?.properties?.tab);
```
We should have unique tab identifiers and this would do the trick. @rawagner wdyt?
I guess getTabCards is also doing something similar so do we really need a new extension ?

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx [hunk 14]
```diff
@@ -50,25 +47,24 @@ const getPluginTabPages = (tabs: DashboardsTab[], cards: DashboardsCard[]): Page
...
-    ],
-    [pluginPages, t],
-  );
+  const allPages = [
+    {
+      href: '',
+      name: t('dashboard~Cluster'),
+      component: ClusterDashboard,
+    },
+  ];
```
Comment (bipuladh) [inline frontend/public/components/dashboard/dashboards-page/dashboards.tsx:67]: Please revert this change as per https://github.com/openshift/console/pull/8496/files#r609376762

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 14 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:67]: got it... will revert the changes...

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx (hunk 13 above)
Comment (SanjalKatiyar) [reply frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:36]: We are adding new OCS tabs without using the extension (in `ocs-dashboards.tsx` file; Not in `plugin.ts` file). These OCS cards will not overlap/interfere with any of the cluster overview tabs.

Or, if that is not the right way, we could add one more property to the extension type: 'Dashboards/Tab', let's say navSection which specifies the nav section to which that dashboard tab belongs to. For example: for cluster overview tabs value will be “cluster” and for OCS specific tabs it will be “storage” and we will add one more filter on this property while displaying the dashboards.
So, the extension will look something like this:
```
{
   type: 'Dashboards/Tab',
   properties: {
     id: 'persistent-storage',
     navSection: 'storage',
     // t('ceph-storage-plugin~Persistent Storage')
     title: '%ceph-storage-plugin~Persistent Storage%',
   },
   flags: {
     required: [FLAG],
   },
 }
```
What do you guys think should be better?

File: frontend/public/components/dashboard/dashboards-page/dashboards.tsx (hunk 12 above)
Comment (SanjalKatiyar) [reply frontend/public/components/dashboard/dashboards-page/dashboards.tsx:51]: ack... I will revert the changes...

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx (hunk 13 above)
Comment (rawagner) [reply frontend/packages/ceph-storage-plugin/src/components/dashboards/ocs-dashboards.tsx:36]: +1 for adding `navSection`

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 11 above)
Comment (SanjalKatiyar) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts:669]: got it... I will address it once I get thumbs up from @yuvalgalanti ... have made rest of the changes...

File: frontend/packages/console-plugin-sdk/src/typings/dashboards.ts [hunk 15]
```diff
@@ -131,6 +131,9 @@ namespace ExtensionProperties {
     /** The tab's ID which will be used as part of href within dashboards page */
     id: string;
 
+    /** NavSection to which the tab belongs to */
+    navSection: string;
```
Comment (rawagner) [inline frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:135]: lets be more strict and specify values which are valid.
```suggestion
    navSection: 'home' | 'storage';
```

File: frontend/packages/ceph-storage-plugin/src/plugin.ts (hunk 11 above)
Comment (SanjalKatiyar) [reply frontend/packages/ceph-storage-plugin/src/plugin.ts:641]: We can't change it to `OCS Overview`... so I am leaving it as it is for now...
